### ISSUE TITLE
Missing `_compat/typing.py` file for `pyproject` setups

### ISSUE THREAD

**Issue by kratsg** [scikit-hep/cookie#126]
I think it's probably more common for the ruff `typing-modules` configuration to be `{package}.typing` rather than with `_compat` in here. So perhaps throwing in `py.typed` and `typing.py` would do the trick: https://github.com/scikit-hep/cookie/blob/df59a9bccbe1d0f52de5042fe21a29e7a6793d8a/%7B%7Bcookiecutter.project_name%7D%7D/pyproject.toml#L260

**Comment by henryiii** [comment]
It should be there: https://github.com/scikit-hep/cookie/blob/main/%7B%7Bcookiecutter.project_name%7D%7D/src/%7B%7Bcookiecutter.project_name%7D%7D/_compat/typing.py - unless it's missing being copied from one of the backends!

`package.typing` is where you put public typing definitions for others to use. `package._compat.typing` is where you'd put backward compatibility shims for `typing_extensions`. These are not public - you will delete them when you drop support for Pythons where they are not already present. The main thing Ruff needs is to know where `Literal` is coming from, which is a backward compatibility shim.

See https://github.com/scikit-build/scikit-build-core/tree/main/src/scikit_build_core/_compat for an example of how you could use `_compat`.

**Comment by kratsg** [comment]
Oh, thanks for the clarification. I really did not understand what was happening here. Maybe I'm missing it (I set up hatch + pyproject defaults and I think I'm not seeing it show up).

> `package.typing` is where you put public typing definitions for others to use. `package._compat.typing` is where you'd put backward compatibility shims for `typing_extensions`. These are not public - you will delete them when you drop support for Pythons where they are not already present.

Ok, I think I understand this distinction. is this a design pattern documented somewhere, or something you're encouraging for scikit-hep developers?

> The main thing Ruff needs is to know where `Literal` is coming from, which is a backward compatibility shim.

How so? Looking at the scikit-build-core example, I see you do something like

```python
from .._compat.builtins import ExceptionGroup
```

so wouldn't Ruff know that `ExceptionGroup` comes from `_compat.builtins`? Or is Ruff not parsing these imports?

**Comment by henryiii** [comment]
> I set up hatch + pyproject defaults and I think I'm not seeing it show up

I recently added testing for the existence of these, so hopefully it's working now if it wasn't before. I _think_ only one of the backends (not hatch) was not picking it up, but I'm not sure.

> Is this a design pattern documented somewhere, or something you're encouraging for scikit-hep developers

It's something I've started doing more of. We don't really have a great spot to make a recommendation like that in the developer docs, at least yet. Most of our packages do have a similar compatibility shim, though names and locations vary (in fact, I'm still doing this in `hist.typing`, but it really needs to move).

> so wouldn't Ruff know that

Only `Literal` is treated specially, as Ruff needs to know that `x: Literal["hi"]` can't be simplified to `x: Literal[hi]`, even with future annotations on. It knows `typing.Literal` and `typing_extensions.Literal`, but not `mypackage.stuff.Literal` unless you tell it where your typing module shim is. It will assume the module indicated here is a re-export shim package.

Other imports, like `ExceptionGroup`, don't matter, just (one or maybe some) typing ones.
